design/volume-group-snapshot.md

```diff
@@ -0,0 +1,605 @@
+    PVC3 --- PVC4
+    PVC3 -- "group: B" --> VGS_B[VGS-B group: B]
+    PVC4 -- "group: B" --> VGS_B
+```
+
+3. Pod Mounts: Pod1 mounts both PVC1 and PVC2, Pod2 mounts PVC1 and PVC3.  
+   Grouping:   
+     Group A: PVC1 and PVC2  
+     Group B: PVC1 and PVC3  
+   ItemBlock: All objects-Pod1, Pod2, PVC1, PVC2, and PVC3, are collected into a single item block.  
+   VolumeGroupSnapshots:  
+     PVC1 and PVC2 (group A) point to the same VGS (VGS (group: A)).   
+     PVC3 (group B) point to a different VGS (VGS (group: B)).  
+```mermaid
+flowchart TD
+    subgraph ItemBlock
+        P1[Pod1]
+        P2[Pod2]
+        PVC1[PVC1 group: A, group: B]
+        PVC2[PVC2 group: A]
+        PVC3[PVC3 group: B]
+    end
+
+    %% Pod mount relationships
+    P1 -->|mounts| PVC1
+    P1 -->|mounts| PVC2
+    P2 -->|mounts| PVC1
+    P2 -->|mounts| PVC3
+
+    %% Grouping for Group A: PVC1 and PVC2 are grouped into VGS_A
+    PVC1 --- PVC2
+    PVC1 -- "Group A" --> VGS_A[VGS Group A]
+    PVC2 -- "Group A" --> VGS_A
+
+    %% Grouping for Group B: PVC3 grouped into VGS_B
+    PVC3 -- "Group B" --> VGS_B[VGS Group B]
+    
+```
+
+#### Updates to CSI PVC plugin:
+  - This is the plugin which creates VolumeSnapshots for the CSI PVCs if required.
+  - We might encounter the following cases:
+    - PVC has VGS label and VGS already exists:
+      - We can check if VGS exists by listing the VGS in the namespace using the backup UUID 
+        along with the label used for the VGS (we will be creating VGS with backup UUID and user specified VGS  group label) and see if the PVC is part of any of the existing VGS for the particular backup operation we are currently in.
+      - If not then create the VGS object with the required labels (in this case backup UUID and the user specified VGS group label) and also add the PVC names (that are included in the VGS) as 
+        annotations. (This will be useful for our VGS restore operation where we would need PVC resource identifiers).
+      - For non-datamover case:
+        - Once VGS is created, add the VGS as an additional item to be included in the backup.
+        - The VGS BIA plugin will add the related VS as additional items later.
+      - For datamover case:
+        - We need to bypass creation of VS but need to create DataUploads for the VS instance that got created due to VGS creation.
+        - Fetch the correct VS instances that got created via VGS and is related to the current PVC.
+        - Finally create DataUpload object for this VS/PVC pair.
+        - Add the DataUpload as an additional item.
+    - PVC has VGS label and VGS exists:
+      - In this case we do not want to re-create VGS object.
+      - The required VS/DU for the PVC already exists as VGS instance exists for the PVC and backup UUID.
+    - PVC does not possess VGS label:
+      - Create VS and follow the existing legacy workflow.
+
+
+#### Add a new VolumeGroupSnapshot(VGS) BIA plugin
```

Comment:
I further considered this, and I think we don't need to backup the VGS/VGSC/VGSClass objects at all, we only need to create VGS to make sure the snapshots are created at the same time, and after that, the will be out of use, especially, during restore.

During restore, we will recreate the VS/VSC, and the most essential information is the snapshot handle. We don't need VGS/VGSC/VGSClass objects.

Reply:
I tried to set up a VolumeGroupSnapshot-enabled environment, but I haven't managed to do that in a GKE cluster yet.  I will continue to do that later.

As a result, I'm still not sure how the VolumeGroupSnapshot works. 
If the VolumeGroupSnapshot feature is enabled in a cluster, are the VolumeSnapshot-related resources still needed to work?
A concrete scenario is, say the user created a VolumeGroupSnapshot, and it referenced an existing VolumeGroupSnapshotClass, then the VolumeSnapshots should be created accordingly, do the VolumeSnapshots still need to reference a VolumeSnapshotClass, or the `VolumeSnapshotClass` parameter can be nil for the VolumeSnapshots?

Reply:
@blackpiglet for restore we can directly use the VG as a source. https://kubernetes.io/blog/2024/12/18/kubernetes-1-32-volume-group-snapshot-beta/#how-to-use-group-snapshot-for-restore-in-kubernetes 
@Lyndon-Li So use VGS only for co-ordinating VS/VSC creation in the backup workflow, thats it ?

Reply:
> So use VGS only for co-ordinating VS/VSC creation in the backup workflow, thats it ?

Yes, correct

Reply:
> for restore we can directly use the VG as a source.

What is the benefit to do this in restore to Velero/Velero users?
As far as I can come out, it simplifies users' operation when there are many volumes in the VG. However, as a backup solution Velero has already automatically managed volumes->volume snapshots->volume snapshot restores. So this won't help Velero.

Otherwise, if we get rid of VGS in restore, we will not need to restore VGS objects to users' env, which may make chaos and confuse users when an accident happens.

Reply:
@shubham-pampattiwar "for restore we can directly use the VG as a source." -- I think you meant VS right?

We don't have much choice here. The API doesn't allow for using the VGS for restore as I understand it -- VGS facilitates creating individual VolumeSnapshots, but when restoring, there's no choice but to individually create PVCs with a VolumeSnapshot source.

Reply:
@shubham-pampattiwar 
I remember the external-snapshotter requires VolumeSnapshot or the VolumeSnapshotContent's referencing VolumeSnapshotClass existing the cluster. That's why the CSI plugin syncs the VolumeSnapshotClass alongside with the CSI backup.